torch/optim/optimizer.py

```diff
@@ -383,9 +384,11 @@ def zero_grad(self, set_to_none: bool = False):
                             else:
                                 p.grad.requires_grad_(False)
                             if (not foreach or p.grad.is_sparse):
-                                p.grad.zero_()
+                                tmp_group.append(p)
```

Comment:
@albanD Thanks for your reply. Here are the updates.
```suggestion
                                if p.grad.device.type == 'cuda':
                                    tmp_group.append(p)
                                else:
                                    p.grad.zero_()
```